### DISCUSSION TITLE
Recursive dataclass return type not supported

### DISCUSSION THREAD

**Discussion by Garrett-R** [fastapi/fastapi#10722] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Example Code

```python
from dataclasses import dataclass
from typing import Self

from fastapi import FastAPI


app = FastAPI()


@dataclass
class Section:
    title: str
    children: list[Self]


@app.get("/recursive")
async def get_recursive_dataclass() -> Section:
    raise Exception()
```

### Description

* With the snippet above, I run `uvicorn main:app --reload`
* I'm getting:
```python
fastapi.exceptions.FastAPIError: Invalid args for response field! Hint: check that <class 'main.Section'> is a valid Pydantic field type. If you are using a return type annotation that is not a valid Pydantic field (e.g. Union[Response, dict, None]) you can disable generating the response model from the type annotation with the path operation decorator parameter response_model=None. Read more: https://fastapi.tiangolo.com/tutorial/response-model/
```

### Operating System

Linux

### Operating System Details

Ubuntu 22.04

### FastAPI Version

0.104.1

### Pydantic Version

2.5.1

### Python Version

3.11.6

**Comment by JavierSanchezCastro** [comment]
Try this:

```python3
from dataclasses import dataclass
from fastapi import FastAPI


app = FastAPI()

@dataclass
class Section:
    title: str
    children: list["Section"]
    
sections = Section(title='title 1', children=[
     Section(title='title 11', children=[]),
     Section(title='title 12', children=[
         Section("another", children=[])
     ])
])

@app.get("/recursive", response_model=None)
async def get_recursive_dataclass() -> Section:
    return sections
```

you can also use

``from __future__ import annotations``

**Comment by Ventura94** [comment]
You just have to add ```response_model=None``` to the decorator, leaving it as follows: 
````python 
from dataclasses import dataclass
from typing import Self

from fastapi import FastAPI


app = FastAPI()


@dataclass
class Section:
    title: str
    children: list[Self]


@app.get("/recursive", response_model=None)
async def get_recursive_dataclass() -> Section:
    raise Exception() 
````

**Reply by Garrett-R** [reply]
That works, thanks!

**Comment by dozed** [comment, marked as answer]
Another solution is to use the `pydantic.dataclasses.dataclass` annotation. With this, the model will also be available in the OpenAPI docs:

```python
from __future__ import annotations

from pydantic.dataclasses import dataclass
from fastapi import FastAPI

app = FastAPI()


@dataclass
class Section:
    title: str
    children: list[Section]


sections = Section(title='title 1', children=[
    Section(title='title 11', children=[]),
    Section(title='title 12', children=[
        Section("another", children=[])
    ])
])


@app.get("/recursive")
async def get_recursive_dataclass() -> Section:
    return sections
```

Side note: a related issue with forward references is: `pydantic.errors.ConfigError: field "..." not yet prepared so type is still a ForwardRef, you might need to call ....update_forward_refs().`

This can solved by invoking `update_forward_refs()` after the definition of all types as shown here:

```python
from __future__ import annotations

from pydantic.dataclasses import dataclass
from fastapi import FastAPI

app = FastAPI()


@dataclass
class Book:
    sections: list[Section]


@dataclass
class Section:
    title: str
    children: list[Section]


Book.__pydantic_model__.update_forward_refs()

sections = Section(title='title 1', children=[
    Section(title='title 11', children=[]),
    Section(title='title 12', children=[
        Section("another", children=[])
    ])
])

book = Book([sections])


@app.get("/recursive")
async def get_recursive_dataclass() -> Book:
    return book
```

**Reply by Garrett-R** [reply]
That's great!  And looks like it will indeed require using `from __future__ import annotations` until [this Pydantic issue](https://github.com/pydantic/pydantic/issues/5992) get solved.

**Comment by rafalkrupinski** [comment]
It's actually a problem with pydantic not handling Self in field type (pydantic/pydantic#5992). ~~@JavierSanchezCastro~~ @dozed  gave a better answer.

**Reply by dozed** [reply]
The issue with his answer is that due to `response_model=None` the OpenAPI docs are missing the return type details.

pydantic fails to generate the schema, due to an error in this line, since `field_type` is of type `ForwardRef`: https://github.com/pydantic/pydantic/blob/main/pydantic/v1/schema.py#L927

**Reply by rafalkrupinski** [reply]
@dozed You're absolutely right. I was focused on the field type - Self vs "Section".

It should be either `"list[Section]"`, or, as you wrote it, `list[Section]` with `from __future__ import annotations`
